### DISCUSSION TITLE
Subcomponents parameter of Module annotation doesn't seem to work.

### DISCUSSION THREAD

**Discussion by kylannjohnson** [ZacSweers/anvil#60] [Q&A] [answered]
This is a small issue, but the solution wasn't entirely obvious so I thought I'd share my experience and see if something could be improved.

It seems like the move to KSP has changed the mechanism of how Subcomponents are linked to their parents.  Moving to this fork required the inner Builder interface to use the Builder interface from its parent interface.  e.g. MergeComponent now requires `@MergeComponent.Builder` instead of just `@Component.Builder`.

In my case, we have an AppComponent and a SubComponent called the InitializedComponent.  They are both defined in the `:app` module like this.

```kotlin
@MergeComponent(AppScope::class)
interface AppComponent {
    fun inject(mainActivity: MainActivity)

    @Component.Builder
    interface Builder {
        @BindsInstance
        fun application(application: Application): Builder
        fun build(): AppComponent
    }
}

@MergeSubcomponent(scope = InitializedScope::class)
interface InitializedComponent : InjectorsComponent {

    @Subcomponent.Factory
    interface Factory {
        fun build(): InitializedComponent
    }
}
```

On the base version of Anvil, we then "link" the two components by creating a module like this and using the `subcomponents` option.

```kotlin
@[Module(subcomponents = [InitializedComponent::class]) ContributesTo(AppScope::class)]
object InitializerModule {

    @Provides
    fun provideInitializer(factory: InitializedComponent.Factory): Initializer {
        return Initializer(factory)
    }
}
```

However, because of the requirement that the Factory (or Builder) must be annotated with the new annotation, we now get this KSP error

```
...Initializer.kt:11: com.dagger.InitializedComponent is not a @Subcomponent or @ProductionSubcomponent
```

The workaround is to add `val factory: InitializedComponent.Factory` to the AppComponent interface.   Should this be in documentation? Or is this just a not-yet-supported feature?

Thanks for working on this!

**Comment by ZacSweers** [comment, marked as answer]
Yeah this is one of the cases described at the bottom of this section: https://github.com/ZacSweers/anvil/blob/main/FORK.md#2-ksp-component-merging

Anywhere that was not processed by anvil before and referenced anvil merged types need to be updated to use the generated `Merged*` class instead. In your case, something like this

```kotlin
@[Module(subcomponents = [MergedInitializedComponent::class]) ContributesTo(AppScope::class)]
object InitializerModule {

    @Provides
    fun provideInitializer(factory: InitializedComponent.Factory): Initializer {
        return Initializer(factory)
    }
}
```

**Reply by kylannjohnson** [reply]
Thanks for pointing out that section.  I reread it and it seems to make sense.  Using the Merged Component did get past the error above.

For what it's worth, the reference to the MergedInitializedComponent in the `@Module` of the InitializerModule wasn't necessary.  Even after removing it everything compiled and worked as before.  The addition of the property in the "parent" interface seemed to be the cause.  Adding it resulted in the creation of a `public abstract class SubcomponentModule` in the MergedInitializedComponent that binds the Factory of the sub component to the parent (in my example AppComponent)
